File: level1/876/middleLinkList.py
class ListNode(object):
	def __init__(self, val=0, next=None, head=None):
		self.val = val
		self.next = None
		self.head = None
	def insert(self, val):
		newNode = ListNode(val)
		if(self.head):
			current = self.head
			while(current.next):
				current = current.next
			current.next = newNode
		else:
			self.head = newNode
	def printList(self):
		current = self.head
		while(current):
			print(current.val, end=' ')
			current = current.next

class Solution(object):
	def middleNode(self, head):
		temp = head
		count = 0
		while (temp != None):
			count += 1
			temp = temp.next
		middle = int(count / 2)
		while (middle):
			middle -= 1
			head = head.next
		return head

# if __name__=='__main__':
# 	LL1 = ListNode()
# 	LL1.insert(2)
# 	LL1.insert(4)
# 	LL1.insert(6)
# 	LL1.insert(8)
# 	print("Original List")
# 	LL1.printList()
# 	print("\nMiddle List")
# 	solution = Solution()
# 	LL2 = solution.middleNode(LL1.head)
# 	LL2.printList()

File: level1/876/middleLinkList.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
	int val;
	struct ListNode *next;
 };
 
int	getListSize(struct ListNode *head)
{
	struct ListNode *temp = head;
	int				count = 0;

	while(temp)
	{
		temp = temp->next;
		count++;
	}
	return(count);
}

struct ListNode* middleNode(struct ListNode* head){
	int ListSize = getListSize(head);
	int	midPoint = ListSize / 2;
	while (midPoint)
	{
		head = head->next;
		midPoint--;
	}
	return (head);
}

int	main(void)
{
	struct ListNode *middle;
	struct ListNode *list1;
	//struct ListNode f = { 6, NULL };
	struct ListNode e = { 5, NULL };
	struct ListNode d = { 4, &e };
	struct ListNode c = { 3, &d };
	struct ListNode b = { 2, &c };
	struct ListNode a = { 1, &b };
 	list1 = &a;

	middle = middleNode(list1);
	while (middle)
	{
		printf("%d ", middle->val);
		middle = middle->next;
	}
	return (0);
}

